### PR TITLE
cmd/logcli: support load config from file

### PR DESCRIPTION
**Pull request by daixiang0**
Support load config from file and move default config into here.

Signed-off-by: Xiang Dai <764524258@qq.com>

### CODE REVIEW

File: cmd/logcli/config.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+	Password string `yaml:"password,omitempty"`
+}
+
+func getConfig(configFile string) *Config {
+	if configFile == "" {
+		home := os.Getenv("HOME")
+		configFile = fmt.Sprintf("%s/%s", home, "logcli-config.yaml")
+	}
+
+	config := Config{"https://log-us.grafana.net", "", ""}
```
Comment (tomwilkie) [inline cmd/logcli/config.go]: Can you use field names in structs please?  Makes the code less fragile when we add more fields.

File: cmd/logcli/config.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+	if err := helpers.LoadConfig(configFile, &config); err != nil {
```
Comment (tomwilkie) [inline cmd/logcli/config.go]: I think I'd prefer if this function returned `(*Config, error)` and the called dealt with logging errors.

File: cmd/logcli/main.go [hunk 3]
```diff
@@ -7,10 +7,13 @@ import (
...
 
 var (
-	app      = kingpin.New("logcli", "A command-line for loki.")
-	addr     = app.Flag("addr", "Server address.").Default("https://log-us.grafana.net").Envar("GRAFANA_ADDR").String()
-	username = app.Flag("username", "Username for HTTP basic auth.").Default("").Envar("GRAFANA_USERNAME").String()
-	password = app.Flag("password", "Password for HTTP basic auth.").Default("").Envar("GRAFANA_PASSWORD").String()
+	app = kingpin.New("logcli", "A command-line for loki.").DefaultEnvars()
+
+	config   = app.Flag("config", "Logcli config.").Default("").String()
+	cfg      = getConfig(*config)
```
Comment (tomwilkie) [inline cmd/logcli/main.go]: I don't see where this is used?  What am I missing?

**Comment by tomwilkie** [comment]
Thanks for the PR @daixiang0! Looks like it need some more work, ping me when you've had a chance to address the feedback.

File: cmd/logcli/main.go (hunk 3 above)
Comment (daixiang0) [reply cmd/logcli/main.go]: cfg value use in the following code.

**Comment by daixiang0** [comment]
@tomwilkie hi, i do some update, happy to get your feedback

**Comment by daixiang0** [comment]
@tomwilkie any input?

**Comment by daixiang0** [comment]
@tomwilkie any input?

File: cmd/logcli/main.go [hunk 4]
```diff
@@ -26,6 +29,9 @@ var (
 )
 
 func main() {
+	if err != nil {
+		// TODO
```
Comment (davkal) [inline cmd/logcli/main.go]: Should exit if config file not present or could not be parsed.

Comment (daixiang0) [reply cmd/logcli/main.go]: All done, please look, thanks.

Comment (daixiang0) [reply cmd/logcli/main.go]: seems lint check has some issue.

**Comment by daixiang0** [comment]
> * [ ]  `getConfig` should return with the results from `LoadConfig`
> * [ ]  move `getConfig` call into `main`
> * [ ]  handle errors and exit

@davkal All done, please review again.

**Comment by davkal** [comment]
Isn't the error logged twice now?

Also, @gouthamve is this good code organization in the `var()` section?

**Comment by daixiang0** [comment]
> Isn't the error logged twice now?

Do you mean no need log in config.go? I have removed that.

File: docs/logcli.md [hunk 5]
```diff
@@ -20,6 +20,12 @@ Common labels: {job="cortex-ops/consul", namespace="cortex-ops"}
 2018-06-25T12:52:09Z {instance="consul-8576459955-pl75w"} 2018/06/25 12:52:09 [INFO] raft: Snapshot to 475409 complete
 2018-06-25T12:52:09Z {instance="consul-8576459955-pl75w"} 2018/06/25 12:52:09 [INFO] raft: Compacting logs from 456973 to 465169
 ```
+If you want to use config stored in file, you need to specify whole config file path with `--config` option, please take `cmd/logcli/logcli-config.yaml` as example.
```
Comment (davkal) [inline docs/logcli.md]: This sentence needs cleaning up: 
"You may use `--config=path/to/file` to load configuration options from a file. For an example file see `cmd/logcli/logcli-config.yaml`.

File: docs/logcli.md [hunk 6]
```diff
@@ -20,6 +20,12 @@ Common labels: {job="cortex-ops/consul", namespace="cortex-ops"}
...
+
+The config values overvide rule is as below:
```
Comment (davkal) [inline docs/logcli.md]: "Configuration values are considered in the following order (lowest to highest):"

File: docs/logcli.md [hunk 7]
```diff
@@ -34,6 +40,8 @@ Flags:
   --addr=""      Server address, need to specify.
   --username=""  Username for HTTP basic auth.
   --password=""  Password for HTTP basic auth.
+  --config="$HOME/logcli-config.yaml"
```
Comment (davkal) [inline docs/logcli.md]: `--config="" Configuration file for logcli`

**Comment by davkal** [comment]
Thank you, we're almost there. Only the docs need cleaning up, then we can merge this.

**Comment by daixiang0** [comment]
> Thank you, we're almost there. Only the docs need cleaning up, then we can merge this.

Done :）

**Comment by davkal** [comment]
Thank you for your contribution @daixiang0 🎉

**Pull request merged**